# Transformers pour la classification de texte

L'objectif de ce TP est d'implémenter une version simplifiée d'un Transformer pour résoudre un problème de classification de texte.

Nous utiliserons comme exemple illustratif une base de données présente dans la librairie ```Keras``` consistant en des critiques de films postées sur le site IMDB, accompagnées d'une note qui a été binarisée pour révéler le caractère positif, ou négatif, de la critique.

In [1]:
import tensorflow as tf
from tensorflow import keras

## Implémentation d'un bloc de base de Transformer


<center><img src="https://drive.google.com/uc?id=1leAVyTZJ2gZ26CFoauMtmu7T4Jo-Crc_" width=200> </center>
<caption><center> Figure 1: Schéma de l'architecture de BERT</center></caption>

La figure ci-dessus présente l'architecture de BERT. Le bloc de base d'un Transformer est composé d'un bloc de *Self-Attention*, d'une couche de ```Layer Normalization```, d'une couche dense et enfin d'une nouvelle couche de ```Layer Normalization```. L'idée de la couche de ```Layer Normalization```est de faciliter le processus d'apprentissage en renormalisant les activations après chaque couche du réseau afin d'éviter que les valeurs ne s'écartent trop de 0.
Prêtez également bien attention aux **couches résiduelles**.

Pour implémenter la *Self-Attention*, vous pouvez utiliser la fonction ```Multi-Head Attention``` (à vous de regarder quels en sont les paramètres dans la documentation).

**Rappel**: Une couche d'Attention *Multi-Head*  se présente sous la forme ci-dessous à gauche, avec le mécanisme d'attention détaillé à droite :


<center>

<img src="https://drive.google.com/uc?id=1UTozEHtsZ3xy61XJqn_Eug-7mn7bFp9m">
<img src="https://drive.google.com/uc?id=1aTttpp1OOasVVZAi3lWwosh68VnBjQnz">
</center>

**D'après vous, combien de paramètres comporte une couche d'attention à 2 têtes, pour un *Embedding* de dimension 32 ?**


Pour implémenter un bloc Transformer, je vous propose dans ce TP d'adopter une notation ressemblant beaucoup à Pytorch. Nous allons instancier une classe *Layer*, et notamment les deux fonctions :    
-  ```__init__()```, un constructeur dans lequel nous initialisons toutes les couches que nous allons utiliser
-  ```call()``` appelé pour la prédiction et qui décrit comment les différentes couches sont appliquées aux entrées.

In [2]:
class TransformerBlock(keras.layers.Layer):
    # embed_dim désigne la dimension des embeddings maintenus à travers les différentes couches,
    # et num_heads le nombre de têtes de la couche d'attention.
    # DANS CETTE FONCTION, ON NE FAIT QUE DEFINIR LES COUCHES
    def __init__(self, embed_dim, num_heads):
        super().__init__()
        # Définition des différentes couches qui composent le bloc
        # Couche d'attention
        self.att = keras.layers.MultiHeadAttention(num_heads, embed_dim)
        # Première couche de Layer Normalization
        self.layernorm1 = keras.layers.LayerNormalization()
        # Couche Dense (Feed-Forward)
        self.ffn = keras.layers.Dense(embed_dim, activation='relu')
        # Deuxième couche de normalisation
        self.layernorm2 = keras.layers.LayerNormalization()

    # DANS CETTE FONCTION, ON APPELLE EXPLICITEMENT LES COUCHES DEFINIES DANS __init__
    # ON PROPAGE DONC LES ENTREES inputs A TRAVERS LES DIFFERENTES COUCHES POUR OBTENIR
    # LA SORTIE
    def call(self, inputs):
        # Application des couches successives aux entrées
        entree = self.att(inputs, inputs)
        s1 = self.layernorm1(entree + inputs)
        s2 = self.ffn(s1)
        s3 = self.layernorm2(s2 + s1)
        return s3

## Implémentation de la double couche d'Embedding

La séquence d'entrée est convertie en *Embedding* de dimension ```embed_dim```.
L'*Embedding* final est constitué de la somme de deux *Embedding*, le premier encodant un mot, et le second encodant la position du mot dans la séquence.

La couche d'*Embedding* de Keras (```layers.Embedding```) est une sorte de table associant à un indice en entrée un vecteur de dimension ```embed_dim```. Chaque coefficient de cette table est en fait un paramètre apprenable.

**D'après vous combien de paramètres contiendrait une couche d'*Embedding* associant un vecteur de dimension 32 à chacun des 20000 mots les plus courants du vocabulaire extrait de la base de données que nous allons utiliser ?
Et combien pour l'*Embedding* qui associe un vecteur de dimension 32 à chaque position d'un séquence de longueur ```maxlen``` ?**

In [3]:
class TokenAndPositionEmbedding(keras.layers.Layer):
    def __init__(self, maxlen, vocab_size, embed_dim):
        super().__init__()
        # Définition des différentes couches qui composent le bloc Embedding
        # Embedding de mot
        self.token_emb = keras.layers.Embedding(vocab_size, embed_dim)
        # Embedding de position
        self.pos_emb = keras.layers.Embedding(maxlen, embed_dim)

    def call(self, x):
        # Calcul de l'embedding à partir de l'entrée x
        # ATTENTION : UTILISER UNIQUEMENT DES FONCTIONS TF POUR CETTE PARTIE
        # Récupération de la longueur de la séquence
        maxlen = tf.shape(x)[1]
        # Création d'un vecteur [0, 1, ..., maxlen] des positions associées aux
        # mots de la séquence (fonction tf.range)
        positions = tf.range(0,maxlen)
        # Calcul des embeddings de position
        positions_emb = self.pos_emb(positions)
        # Calcul des embeddings de mot
        words_emb = self.token_emb(x)
        return tf.math.add(positions_emb,words_emb)

## Préparation de la base de données

In [4]:
# Taille du vocabulaire considéré (on ne conserve que les 20000 mots les plus courants)
vocab_size = 20000
# Taille maximale de la séquence considérée (on ne conserve que les 200 premiers mots de chaque commentaire)
maxlen = 200

# Chargement des données de la base IMDB
(x_train, y_train), (x_val, y_val) = keras.datasets.imdb.load_data(num_words=vocab_size)

print(len(x_train), "séquences d'apprentissage")
print(len(x_val), "séquences de validation")

# Padding des séquences : ajout de "0" pour compléter les séquences trop courtes
x_train = keras.preprocessing.sequence.pad_sequences(x_train, maxlen=maxlen)
x_val = keras.preprocessing.sequence.pad_sequences(x_val, maxlen=maxlen)

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
25000 séquences d'apprentissage
25000 séquences de validation


## Création du modèle

Pour assembler le modèle final, il faut, partant d'une séquence de longueur ```maxlen```, calculer les Embedding puis les fournir en entrée d'une série de blocs Transformer. Pour ce TP, **commencez par ne mettre qu'un seul bloc Transformer**. Vous pourrez en ajouter plus tard si vous le souhaitez.

Pour construire la tête de projection du réseau, vous pouvez moyenner les activations en sortie du bloc Transformer par élément de la séquence grâce à un *Global Average Pooling* (1D !), à relier à une couche dense (par exemple comportant 20 neurones) et enfin à la couche de sortie du réseau.

In [5]:
embed_dim = 32  # Dimension de l'embedding pour chaque mot
num_heads = 2  # Nombre de têtes d'attention

# A COMPLETER
inputs = keras.layers.Input(shape=(maxlen,))
embedding = TokenAndPositionEmbedding(maxlen,vocab_size, embed_dim)(inputs)
transformer = TransformerBlock(embed_dim, num_heads)(embedding)
avgpooling = keras.layers.GlobalAveragePooling1D()(transformer)
dense = keras.layers.Dense(20, activation='relu')(avgpooling)
outputs = keras.layers.Dense(1,activation='sigmoid')(dense)

model = keras.Model(inputs=inputs, outputs=outputs)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)             │ (None, 200)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ token_and_position_embedding         │ (None, 200, 32)             │         646,400 │
│ (TokenAndPositionEmbedding)          │                             │                 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ transformer_block (TransformerBlock) │ (None, 200, 32)             │           9,600 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ global_average_pooling1d             │ (None, 32)                  │               0 │
│ (GlobalAveragePooling1D)             │                             │                 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 20)                  │             660 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              21 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 656,681 (2.51 MB)

 Trainable params: 656,681 (2.51 MB)

 Non-trainable params: 0 (0.00 B)

Enfin vous pouvez lancer l'apprentissage, avec par exemple l'optimiseur Adam. Inutile de lancer de trop nombreuses *epochs*, le réseau sur-apprend très vite !

In [6]:
# A COMPLETER
model.compile(
    optimizer="adam", loss="binary_crossentropy", metrics=["accuracy"]
)
history = model.fit(
    x_train, y_train, batch_size=32, epochs=5, validation_data=(x_val, y_val)
)

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 27s 18ms/step - accuracy: 0.7600 - loss: 0.4716 - val_accuracy: 0.8778 - val_loss: 0.2861
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9364 - loss: 0.1697 - val_accuracy: 0.8707 - val_loss: 0.3237
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.9675 - loss: 0.0991 - val_accuracy: 0.8564 - val_loss: 0.3921
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 6s 8ms/step - accuracy: 0.9828 - loss: 0.0608 - val_accuracy: 0.8445 - val_loss: 0.5050
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.9881 - loss: 0.0428 - val_accuracy: 0.8387 - val_loss: 0.5881


In [7]:
y_pred = model.predict(x_train)
print(y_pred[0])
print(x_train[0])

782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step
[0.9997049]
[    5    25   100    43   838   112    50   670     2     9    35   480
   284     5   150     4   172   112   167     2   336   385    39     4
   172  4536  1111    17   546    38    13   447     4   192    50    16
     6   147  2025    19    14    22     4  1920  4613   469     4    22
    71    87    12    16    43   530    38    76    15    13  1247     4
    22    17   515    17    12    16   626    18 19193     5    62   386
    12     8   316     8   106     5     4  2223  5244    16   480    66
  3785    33     4   130    12    16    38   619     5    25   124    51
    36   135    48    25  1415    33     6    22    12   215    28    77
    52     5    14   407    16    82 10311     8     4   107   117  5952
    15   256     4     2     7  3766     5   723    36    71    43   530
   476    26   400   317    46     7     4 12118  1029    13   104    88
     4   381    15   297    98    32  2071    56    26   141     6   19

# Utilisation d'un modèle pré-entraîné

L'implémentation d'un Transformer *from scratch* comme dans la section précédente n'est en fait pas réellement recommandée. Il est souvent bien plus pertinent d'utiliser un modèle pré-entraîné et de faire du transfert d'apprentissage.

Dans cette 2nde partie, nous allons utiliser des modèles fournis par HuggingFace pour appréhender les différentes parties d'un système de classification de texte, puis pour résoudre le problème de la partie précédente d'une manière plus satisfaisante.

## Tokenization

La tokenization désigne l'opération de découpage du texte initial en une séquence de *tokens*, c'est-à-dire d'unités indivisibles de texte. Dans la partie précédente, nous avons implicitement utilisé une simple Tokenization au niveau mot.

Pour jouer un peu avec ce concept, chargeons le Tokenizer du modèle CAMEMBERT. CAMEMBERT est une version de BERT entraînée sur des données uniquement en français. Le Tokenizer a donc été préparé uniquement pour des mots en français.




In [8]:
from transformers import CamembertTokenizer

# Charger le tokenizer CamemBERT
tokenizer = CamembertTokenizer.from_pretrained("camembert-base")

# Phrase en français
sentence = "Ce sont un baobab et un sirop à la fraise."

# Tokenisation de la phrase
encoded = tokenizer(sentence, return_tensors="tf")

# Afficher les IDs des tokens
token_ids = encoded["input_ids"].numpy()[0]  # Convertir en numpy pour affichage
print("IDs des tokens :", token_ids)

# Associer chaque ID au mot correspondant
tokens = [tokenizer.decode([token_id]) for token_id in token_ids]

# Afficher les tokens avec leur ID
print("\nCorrespondance ID -> Token :")
for token_id, token in zip(token_ids, tokens):
    print(f"{token_id} -> {token}")


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/811k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.40M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/508 [00:00<?, ?B/s]

IDs des tokens : [    5   148    56    23  3498   189 16328    14    23 13502    15    13
   767    35     9     6]

Correspondance ID -> Token :
5 -> <s>
148 -> Ce
56 -> sont
23 -> un
3498 -> ba
189 -> o
16328 -> bab
14 -> et
23 -> un
13502 -> sirop
15 -> à
13 -> la
767 -> frais
35 -> e
9 -> .
6 -> </s>


Sur cet exemple, on a l'impression que la Tokenization de CAMEMBERT a été également réalisée au niveau mot. On repère les tokens spéciaux \<s\> (pour démarrer une phrase) et \</s\> (pour la terminer). *Notez ici pour la suite que cette phrase compte 10 tokens.*

En réalité, le Tokenizer est hybride : il décompose les phrases en mots si ceux-ci sont très communs, mais aussi en syllabes ou même en simples caractères.

**Travail à faire : Essayez de trouver des exemples de phrases pour faire apparaître des tokens qui ne sont pas des mots.**

## Complétion de texte masqué par CAMEMBERT

Les modèles de type BERT, comme CAMEMBERT, sont pré-entraînés sur de vastes corpus de textes de manière non supervisée. La méthode d'entraînement la plus commune est de leur faire compléter des morceaux de phrase qui ont été masqués à l'aide d'un token spécial.


Il est intéressant de tester la capacité du modèle à compléter un texte masqué en appliquant la procédure suivante :

In [27]:
from transformers import TFCamembertForMaskedLM
import tensorflow as tf

# Chargement du modèle Camembert
model = TFCamembertForMaskedLM.from_pretrained("camembert-base")

# Phrase en français masquée
sentence = "C'est l'histoire d'un <mask>, il est mort."

# Tokenisation de la phrase
encoded = tokenizer(sentence, return_tensors="tf")

# Prédiction du modèle sur la phrase
outputs = model(encoded)

# Affichage de la dimension du tenseur prédit
print(outputs.logits.shape)


All PyTorch model weights were used when initializing TFCamembertForMaskedLM.

All the weights of TFCamembertForMaskedLM were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFCamembertForMaskedLM for predictions without further training.


(1, 18, 32005)


Le tenseur prédit par le modèle CAMEMBERT sur la phrase d'exemple est de taille $1 \times 10 \times 32005$ :    
- le 1 correspond à la taille du batch (ici simplement 1)
- le 10 correspond au nombre de tokens de la séquence, c'est la valeur que nous avions trouvée précédemment. Le symbole \<mask\> compte pour un token.
- La veleur 32005 correspond au nombre de tokens de la tokenization de CAMEMBERT. On obtient ainsi une distribution de probabilité sur tous ces tokens.

Pour prédire le mot masqué, il suffit de retrouver parmi les 10 tokens, l'indice du token \<mask\> et de regarder la distribution de probabilité prédite par le modèle. On peut ainsi afficher les tokens ayant la probabilité la plus élevée de compléter le trou dans la phrase.

In [28]:
### A COMPLETER
# Récupérer l'index du token masqué
mask_index = tf.where(encoded["input_ids"] == tokenizer.mask_token_id).numpy().flatten()[1]

# Obtenir les logits des prédictions pour le token masqué
logits = outputs.logits[0,mask_index,:]

# Il faut appliquer la fonction softmax pour obtenir une distribution de probabilité
probs = tf.math.softmax(logits)

# On collecte les 5 prédictions les plus probables, ainsi que leur probabilité
top_5 = tf.math.top_k(probs, k=5)

# Associer chaque token ID à son mot et sa probabilité
predictions = [(tokenizer.decode([idx]).strip(), prob) for idx, prob in zip(top_5.indices.numpy(), top_5.values.numpy())]

# Afficher les mots prédits avec leurs probabilités
print("Mots prédits avec probabilités :")
for word, prob in predictions:
    print(f"{word}: {prob:.4f}")

Mots prédits avec probabilités :
homme: 0.1690
écrivain: 0.0365
ami: 0.0352
personnage: 0.0321
musicien: 0.0243


## Transfert d'apprentissage d'un modèle BERT pré-entraîné

Revenons maintenant au problème initial de classification de reviews IMDB.

Nous devons d'abord résoudre un petit problème : la base de données fournie par Keras ne contient que les identifiants des mots écrits dans les reviews, afin de s'abstraire des problématiques de tokenization.

Pour pouvoir utiliser le Tokenizer du modèle que nous allons choisir, il nous faut disposer des phrases originelles. Nous allons adopter une méthode un peu imparfaite et nous allons reconstruire les reviews en utilisant le dictionnaire (index vers mot) fourni par Keras. Malheureusement, comme nous avons choisi de ne conserver que les 20 000 mots les plus courants, il manquera certains mots !

In [29]:
# Récupération du dictionnaire mot->index
word_index = keras.datasets.imdb.get_word_index()

# Insertion des tokens rajoutés par Keras
word_index = {word: (index + 3) for word, index in word_index.items()}
word_index["<PAD>"] = 0   # token de padding
word_index["<START>"] = 1  # token de début de review
word_index["<UNK>"] = 2    # Token pour les mots inconnus
word_index["<UNUSED>"] = 3
print(word_index)

# Inversion du dictionnaire : index->mot
reverse_word_index = dict([(value, key) for (key, value) in word_index.items()])

print(reverse_word_index)

Output hidden; open in https://colab.research.google.com to view.

Nous pouvons maintenant, à l'aide du dictionnaire, retrouver les phrases initiales de la base de données :



In [30]:
def indices_to_text(indices):
    return " ".join(reverse_word_index.get(i, "<UNK>") for i in indices)

# Affichage de la première review de la base d'entraînement
decoded_review = indices_to_text(x_train[0])

print(decoded_review)

and you could just imagine being there robert <UNK> is an amazing actor and now the same being director <UNK> father came from the same scottish island as myself so i loved the fact there was a real connection with this film the witty remarks throughout the film were great it was just brilliant so much that i bought the film as soon as it was released for retail and would recommend it to everyone to watch and the fly fishing was amazing really cried at the end it was so sad and you know what they say if you cry at a film it must have been good and this definitely was also congratulations to the two little boy's that played the <UNK> of norman and paul they were just brilliant children are often left out of the praising list i think because the stars that play them all grown up are such a big profile for the whole film but these children are amazing and should be praised for what they have done don't you think the whole story was so lovely because it was true and was someone's life afte

On observe en effet qu'il manque des mots !

Si nous chargeons maintenant le Tokenizer du modèle BERT, nous pouvons ainsi observer la version tokenizée de la première review de la base d'apprentissage. Vous comprenez donc que le processus que nous venons d'appliquer nous a permis de retrouver la base initiale depuis la Tokenization de Keras, afin de pouvoir préparer cette base à la classification par BERT en lui appliquant le Tokenizer correspondant !

In [31]:
from transformers import AutoTokenizer

# Chargement du Tokenizer du modèle ALBERT
tokenizer = AutoTokenizer.from_pretrained("albert-base-v2")

# Tokenization de la première review du dataset
tokenizer(decoded_review, return_tensors="np", padding=True)

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/684 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/760k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

{'input_ids': array([[    2,    17,    42,   110,   114,  4382,   142,    80,   822,
           13,     1,    25,    40,  5289,  1574,    17,   130,    14,
          205,   142,   559,    13,     1,   321,   281,    37,    14,
          205,  2509,   553,    28,   992,    86,    31,  2199,    14,
          837,    80,    23,    21,   683,  2760,    29,    48,   171,
           14, 27468, 13986,   892,    14,   171,    46,   374,    32,
           23,   114,  7472,    86,   212,    30,    31,  2448,    14,
          171,    28,   651,    28,    32,    23,   261,    26,  5881,
           17,    83, 12360,    32,    20,  1266,    20,  1455,    17,
           14,  2855,  4534,    23,  5289,   510,  5653,    35,    14,
          241,    32,    23,    86,  2950,    17,    42,   143,    98,
           59,   395,   100,    42,  3687,    35,    21,   171,    32,
          491,    57,    74,   254,    17,    48,  4939,    23,    67,
           13, 24807,    20,    14,    81,   265,   883,    22,

Maintenant que cette preuve de concept fonctionne nous pouvons appliquer le même processus à l'ensemble de la base d'apprentissage et de la base de validation :

In [ ]:
# On convertit maintenant l'ensemble des reviews en texte
texts_train = [indices_to_text(review) for review in x_train]
texts_val = [indices_to_text(review) for review in x_val]

# On applique le Tokenizer du modèle ALBERT sur toutes les reviews
x_train_bert = tokenizer(texts_train, return_tensors="np", padding=True)
x_val_bert = tokenizer(texts_val, return_tensors="np", padding=True)

# Enfin, on prépare les tenseurs pour l'apprentissage en les passant au bon format tensorflow
x_train_bert_tensors = {key: tf.convert_to_tensor(value) for key, value in x_train_bert.items()}
x_val_bert_tensors = {key: tf.convert_to_tensor(value) for key, value in x_val_bert.items()}

Nous pouvons maintenant lancer le transfert d'apprentissage du modèle ALBERT, qui est une version plus compacte de BERT. Attention cette dernière cellule est longue ! Cette partie est plus pour vous donner un exemple de fine-tuning mais de nombreux tutorials existent pour l'accélérer (par exemple, en utilisant des modèles quantifiés)

In [ ]:
from transformers import TFAutoModelForSequenceClassification
from transformers import AdamWeightDecay

# Load and compile our model
model = TFAutoModelForSequenceClassification.from_pretrained("albert-base-v2", num_labels=2)
model.summary()
# Lower learning rates are often better for fine-tuning transformers
optimizer = AdamWeightDecay(learning_rate=3e-5)
model.compile(optimizer=optimizer)  # No loss argument!

model.fit(x_train_bert_tensors, y_train, batch_size=8, validation_data=(x_val_bert_tensors, y_val))